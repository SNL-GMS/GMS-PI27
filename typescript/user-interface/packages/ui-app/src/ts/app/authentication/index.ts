import { ianAuthenticator } from './authentication';
import type { Authenticator } from './types';

export const authenticator: Authenticator = ianAuthenticator;
