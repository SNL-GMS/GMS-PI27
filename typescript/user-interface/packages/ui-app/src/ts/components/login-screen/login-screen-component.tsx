import React from 'react';
import { useParams, useSearchParams } from 'react-router-dom';

import { LoginScreenPanel } from './login-screen-panel';
import type { LoginScreenProps } from './types';

/**
 * Creates a functional login screen wrapper around the LoginScreenPanel component.
 * Builds the `from` object containing the path and search parameters
 */
export function LoginScreenComponent(props: LoginScreenProps) {
  const [searchParams] = useSearchParams();
  const { redirectUrl } = useParams();

  // eslint-disable-next-line react/jsx-props-no-spreading
  return <LoginScreenPanel {...props} searchParams={searchParams} redirectPath={redirectUrl} />;
}
