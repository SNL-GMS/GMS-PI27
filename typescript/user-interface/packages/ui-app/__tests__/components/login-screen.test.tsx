/* eslint-disable react/jsx-props-no-spreading */
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import React from 'react';
import type { To } from 'react-router';
import { createSearchParams } from 'react-router-dom';

import { authenticator } from '~app/authentication';
import { LoginScreenPanel } from '~components/login-screen/login-screen-panel';
import type { LoginScreenReduxProps } from '~components/login-screen/types';

jest.mock('react-router', () => {
  const actualRouter = jest.requireActual('react-router');
  return {
    ...actualRouter,
    Navigate: ({ to }: { to: To }) => {
      // to.search is a string. We just validated that. There's something funny going on
      // with eslint thinking that it is an unbound method
      // eslint-disable-next-line @typescript-eslint/unbound-method
      if (typeof to.search === 'string') return <div>{to.search}</div>;
      return 'fail';
    }
  };
});

const reduxProps: LoginScreenReduxProps = {
  redirectPath: '',
  authenticated: false,
  authenticationCheckComplete: true,
  failedToConnect: false,
  setAppAuthenticationStatus: jest.fn()
};

describe('Login screen', () => {
  it('should be defined', () => {
    expect(LoginScreenPanel).toBeDefined();
  });

  it('failed to connect should return non ideal state, No connection to server', () => {
    const failProps = { ...reduxProps, failedToConnect: true, authenticationCheckComplete: false };
    const loginScreen = render(
      <LoginScreenPanel searchParams={undefined} authenticator={authenticator} {...failProps} />
    );
    const loginScreenNonIdealStateNoConnection = loginScreen.queryByText('No connection to server');
    expect(loginScreen).toMatchSnapshot();
    expect(loginScreenNonIdealStateNoConnection).not.toBeNull();
  });

  it('authentication check in progress shows spinner, attempts to login', () => {
    const failProps = { ...reduxProps, authenticationCheckComplete: false };

    const loginScreen: any = render(
      <LoginScreenPanel searchParams={undefined} authenticator={authenticator} {...failProps} />
    );
    const statusText = loginScreen.queryByText('Attempting to login...');
    expect(loginScreen).toMatchSnapshot();
    expect(statusText).not.toBeNull();
  });

  it('Connected, authentication check complete, and not authenticated should return login page', () => {
    const successProps = { ...reduxProps };

    const loginScreen: any = render(
      <LoginScreenPanel searchParams={undefined} authenticator={authenticator} {...successProps} />
    );
    const loginButton = loginScreen.getByRole('button', { name: 'Login' });
    expect(loginButton.getAttribute('data-cy')).toBe('login-btn');
    expect(loginScreen).toMatchSnapshot();
  });

  it('user authenticates, the app should try to update authentication status', async () => {
    const setAppAuthenticationStatus = jest.fn();
    const loginScreen: any = render(
      <LoginScreenPanel
        searchParams={undefined}
        authenticator={authenticator}
        {...reduxProps}
        setAppAuthenticationStatus={user => {
          setAppAuthenticationStatus(user);
        }}
      />
    );
    const usernameField = loginScreen.getByTestId('username-input');
    const loginButton = loginScreen.getByRole('button', { name: 'Login' });
    expect(usernameField).toBeDefined();
    expect(usernameField.value).toMatch('');
    fireEvent.change(usernameField, { target: { value: 'testUser' } });
    expect(usernameField.value).toMatch('testUser');
    fireEvent.click(loginButton);
    await waitFor(() => expect(setAppAuthenticationStatus).toHaveBeenCalled());
  });

  it('Success preserves url search parameters', async () => {
    const successProps = { ...reduxProps, authenticated: true, redirectPath: '/logged-in' };

    render(
      <LoginScreenPanel
        searchParams={createSearchParams({ layout: 'TestLayout' })}
        authenticator={authenticator}
        {...successProps}
      />
    );
    await waitFor(() => expect(screen.getByText('?layout=TestLayout')).toBeDefined(), {
      timeout: 1000
    });
  });
});
