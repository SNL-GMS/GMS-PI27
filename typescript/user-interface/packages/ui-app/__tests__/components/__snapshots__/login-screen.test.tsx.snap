// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login screen Connected, authentication check complete, and not authenticated should return login page 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="bp5-dark login-screen-body"
      >
        <div
          class="login-container"
        >
          <form>
            <img
              alt=""
              id="GMSLogo"
              src="test-file-stub"
              width="370"
            />
            <div
              class="login-legal"
            />
            <div
              class="login-row user"
            >
              <div
                class="login-label"
              >
                Username:
              </div>
              <div
                class="login-input"
              >
                <div
                  class="bp5-input-group login-input"
                >
                  <input
                    class="bp5-input"
                    data-cy="username-input"
                    data-testid="username-input"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
            <div
              class="login-row password"
            >
              <div
                class="login-label"
              >
                Password:
              </div>
              <div
                class="login-input"
              >
                <div
                  class="bp5-input-group bp5-disabled login-input"
                >
                  <input
                    aria-disabled="true"
                    autocomplete="off"
                    class="bp5-input"
                    disabled=""
                    type="password"
                    value=""
                  />
                </div>
              </div>
            </div>
            <div
              class="login-row login-button"
            >
              <button
                class="bp5-button"
                data-cy="login-btn"
                type="button"
              >
                <span
                  class="bp5-button-text"
                >
                  Login
                </span>
              </button>
            </div>
          </form>
        </div>
      </div>
      <div
        class="Toastify"
      />
    </div>
  </body>,
  "container": <div>
    <div
      class="bp5-dark login-screen-body"
    >
      <div
        class="login-container"
      >
        <form>
          <img
            alt=""
            id="GMSLogo"
            src="test-file-stub"
            width="370"
          />
          <div
            class="login-legal"
          />
          <div
            class="login-row user"
          >
            <div
              class="login-label"
            >
              Username:
            </div>
            <div
              class="login-input"
            >
              <div
                class="bp5-input-group login-input"
              >
                <input
                  class="bp5-input"
                  data-cy="username-input"
                  data-testid="username-input"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            class="login-row password"
          >
            <div
              class="login-label"
            >
              Password:
            </div>
            <div
              class="login-input"
            >
              <div
                class="bp5-input-group bp5-disabled login-input"
              >
                <input
                  aria-disabled="true"
                  autocomplete="off"
                  class="bp5-input"
                  disabled=""
                  type="password"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            class="login-row login-button"
          >
            <button
              class="bp5-button"
              data-cy="login-btn"
              type="button"
            >
              <span
                class="bp5-button-text"
              >
                Login
              </span>
            </button>
          </div>
        </form>
      </div>
    </div>
    <div
      class="Toastify"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Login screen authentication check in progress shows spinner, attempts to login 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="bp5-non-ideal-state bp5-non-ideal-state-vertical"
      >
        <div
          class="bp5-non-ideal-state-text"
        >
          <h4
            class="bp5-heading"
          >
            Checking login
          </h4>
          <div>
            Attempting to login...
          </div>
        </div>
        <div
          aria-label="loading"
          aria-valuemax="100"
          aria-valuemin="0"
          class="bp5-spinner bp5-intent-primary"
          role="progressbar"
        >
          <div
            class="bp5-spinner-animation"
          >
            <svg
              height="50"
              stroke-width="8.00"
              viewBox="1.00 1.00 98.00 98.00"
              width="50"
            >
              <path
                class="bp5-spinner-track"
                d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
              />
              <path
                class="bp5-spinner-head"
                d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
                pathLength="280"
                stroke-dasharray="280 280"
                stroke-dashoffset="210"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="bp5-non-ideal-state bp5-non-ideal-state-vertical"
    >
      <div
        class="bp5-non-ideal-state-text"
      >
        <h4
          class="bp5-heading"
        >
          Checking login
        </h4>
        <div>
          Attempting to login...
        </div>
      </div>
      <div
        aria-label="loading"
        aria-valuemax="100"
        aria-valuemin="0"
        class="bp5-spinner bp5-intent-primary"
        role="progressbar"
      >
        <div
          class="bp5-spinner-animation"
        >
          <svg
            height="50"
            stroke-width="8.00"
            viewBox="1.00 1.00 98.00 98.00"
            width="50"
          >
            <path
              class="bp5-spinner-track"
              d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
            />
            <path
              class="bp5-spinner-head"
              d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
              pathLength="280"
              stroke-dasharray="280 280"
              stroke-dashoffset="210"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Login screen failed to connect should return non ideal state, No connection to server 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="bp5-non-ideal-state bp5-non-ideal-state-vertical"
      >
        <div
          class="bp5-non-ideal-state-visual"
          style="font-size: 48px; line-height: 48px;"
        >
          <span
            aria-hidden="true"
            class="bp5-icon bp5-icon-error"
            data-icon="error"
            tabindex="-1"
          />
        </div>
        <div
          class="bp5-non-ideal-state-text"
        >
          <h4
            class="bp5-heading"
          >
            No connection to server
          </h4>
          <div>
            Attempting to connect...
          </div>
        </div>
        <div
          aria-label="loading"
          aria-valuemax="100"
          aria-valuemin="0"
          class="bp5-spinner bp5-intent-danger"
          role="progressbar"
        >
          <div
            class="bp5-spinner-animation"
          >
            <svg
              height="50"
              stroke-width="8.00"
              viewBox="1.00 1.00 98.00 98.00"
              width="50"
            >
              <path
                class="bp5-spinner-track"
                d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
              />
              <path
                class="bp5-spinner-head"
                d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
                pathLength="280"
                stroke-dasharray="280 280"
                stroke-dashoffset="210"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="bp5-non-ideal-state bp5-non-ideal-state-vertical"
    >
      <div
        class="bp5-non-ideal-state-visual"
        style="font-size: 48px; line-height: 48px;"
      >
        <span
          aria-hidden="true"
          class="bp5-icon bp5-icon-error"
          data-icon="error"
          tabindex="-1"
        />
      </div>
      <div
        class="bp5-non-ideal-state-text"
      >
        <h4
          class="bp5-heading"
        >
          No connection to server
        </h4>
        <div>
          Attempting to connect...
        </div>
      </div>
      <div
        aria-label="loading"
        aria-valuemax="100"
        aria-valuemin="0"
        class="bp5-spinner bp5-intent-danger"
        role="progressbar"
      >
        <div
          class="bp5-spinner-animation"
        >
          <svg
            height="50"
            stroke-width="8.00"
            viewBox="1.00 1.00 98.00 98.00"
            width="50"
          >
            <path
              class="bp5-spinner-track"
              d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
            />
            <path
              class="bp5-spinner-head"
              d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
              pathLength="280"
              stroke-dasharray="280 280"
              stroke-dashoffset="210"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
